h/services/bulk_annotation.py

```diff
@@ -0,0 +1,190 @@
+from typing import List, Union
+
+import sqlalchemy as sa
+from sqlalchemy.engine import Row
+from sqlalchemy.orm import Session
+from sqlalchemy.sql import Select
+
+from h.models import Annotation, AnnotationModeration, Group, User
+
+
+class BadDateFilter(ValueError):
+    """There is something wrong with the date filter provided."""
+
+
+class BadFieldSpec(ValueError):
+    """There is something wrong with the field you have specified."""
+
+
+def date_match(column, spec: dict):
+    """
+    Get an SQL comparator for a date column based on dict spec.
+
+    The dict can contain operators as keys and dates as values as per the
+    following complete (but nonsensical) filter:
+
+        {
+            "gt": "2012-11-30",
+            "gte": "2012-11-30",
+            "lt": "2012-11-30",
+            "lte": "2012-11-30",
+            "eq": "2012-11-30",
+            "ne": "2012-11-30",
+        }
+
+    :raises BadDateFilter: For unrecognised operators or no spec
+    """
+    if not spec:
+        raise BadDateFilter(f"No spec given to filter '{column}' on")
+
+    clauses = []
+
+    for op_key, value in spec.items():
+        if op_key == "gt":
+            clauses.append(column > value)
+        elif op_key == "gte":
+            clauses.append(column >= value)
+        elif op_key == "lt":
+            clauses.append(column < value)
+        elif op_key == "lte":
+            clauses.append(column <= value)
+        elif op_key == "eq":
+            clauses.append(column == value)
+        elif op_key == "ne":
+            clauses.append(column != value)
+        else:
+            raise BadDateFilter(f"Unknown date filter operator: {op_key}")
+
+    return sa.and_(*clauses)
+
+
+class BulkAnnotationService:
+    """A service for retrieving annotations for many users at once."""
+
+    # Aliases to distinguish between different types of user
+    _AUTHOR = sa.orm.aliased(User, name="author")
+    _AUDIENCE = sa.orm.aliased(User, name="audience")
+
+    # Acceptable values to pass in as `fields`
+    _FIELDS = {
+        "author.username": _AUTHOR.username,
+        "group.authority_provided_id": Group.authority_provided_id,
+    }
```

Comment:
Striking a slight middle ground between hard-coding and not. We're only coding for the two fields we are interested in, and there's no attempt to map the "author" and "group" part separately from the fields, which we totally could do if we wanted to be more general.

However, in this formulation, you can ask for either field alone, or together, and it's decently easy to add more fields. We are relying on the fact that the fields we are asking for are already available on objects we are joining across.